File: src/cmcandy/solutions/_0018.java
package cmcandy.solutions;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by CHEN on 2020/4/20.
 */
public class _0018 {
    static class Solution {
        public List<List<Integer>> fourSum(int[] nums, int target) {

            //排序
            Arrays.sort(nums);
            List<List<Integer>> ans = new ArrayList( );
            int right, left;
            for (int j = 0; j < nums.length - 3; j++) {
                if ((j >= 1 && nums[j - 1] == nums[j])) continue;
//                if (nums[j] > target) break;
                for (int i = j + 1; i < nums.length - 2; i++) {
                    if ((i > j + 1 && nums[i - 1] == nums[i]))
                        continue;
                    right = nums.length - 1;
                    left = i + 1;
                    while (right > left) {
                        if (nums[j] + nums[right] + nums[i] + nums[left] > target) {
                            right--;
                        } else if (nums[j] + nums[right] + nums[i] + nums[left] < target) {
                            left++;
                        } else if (nums[j] + nums[right] + nums[i] + nums[left] == target) {
                            //可以匹配
                            ans.add(Arrays.asList(nums[j], nums[i], nums[left], nums[right]));
                            while (left < right && nums[left] == nums[left + 1]) left++; // 去重
                            while (left < right && nums[right] == nums[right - 1]) right--; // 去重
                            right--;
                            left++;
                        }
                    }
                }
            }
            return ans;

        }

    }

    public static void main(String[] args) {
        //

        Solution solution = new Solution( );
        System.out.println(solution.fourSum(new int[]{1,-2,-5,-4,-3,3,3,5}, -11));
    }
}

File: src/cmcandy/Python_language_Answers/_0018.py
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        length = len(nums)
        # 偷了个懒用set去重
        res = set()
        # res = []
        for fir in range(0, length - 3):
            if not (fir > 0 and nums[fir - 1] == nums[fir]):
                for sec in range(fir + 1, length - 2):
                    if not (sec > fir + 1 and nums[sec - 1] == nums[sec]):
                        left = sec + 1
                        right = length - 1
                        # 双指针左右逼近
                        while left < right:
                            if nums[fir] + nums[sec] + nums[left] + nums[
                                    right] > target:
                                right -= 1
                            elif nums[fir] + nums[sec] + nums[left] + nums[
                                    right] < target:
                                left += 1
                            else:
                                res.add((nums[fir], nums[sec], nums[left],
                                         nums[right]))
                                left += 1

        return list(res)